### PR TITLE
Optionally print OM created lines

### PR DESCRIPTION
**Pull request by ArthurSens**
This PR adds an extra option to print _created lines when OpenMetrics is the chosen exposition format. It's implemented following the conditional option added in https://github.com/prometheus/common/pull/504.

### CODE REVIEW

File: prometheus/promhttp/http.go
Comment (bwplotka) [inline prometheus/promhttp/http.go:380]:
```suggestion
	EnableOpenMetricsCreatedMetrics bool
```

Nice, let's do extensive docs here on how to use it and what are the consequences.

Comment (bwplotka) [reply prometheus/promhttp/http.go:380]: Also I still think this is dangerous option due to name conflicts and size of scrape 🙈

Comment (ArthurSens) [reply prometheus/promhttp/http.go]: > Also I still think this is dangerous option due to name conflicts and size of scrape 🙈

Not sure what can we do about that besides documenting it properly 😬

File: prometheus/promhttp/http.go
Comment (bwplotka) [inline prometheus/promhttp/http.go:393]: Nice. I wonder if we want to keep this or perhaps deprecate "EnableOpenMetrics" and have "OpenMetrics: *OpenMetricOpts" to add more stuff. We might need this to control:

* Created metrics in text
* Version
* ?

just an idea to not have trillion bool flags.

**Review by bwplotka — APPROVED** [review]
Thanks! One minor nit, otherwise well done!

File: prometheus/promhttp/http.go
Comment (ArthurSens) [reply prometheus/promhttp/http.go:393]: Good idea 👍

File: prometheus/promhttp/http.go
Comment (bwplotka) [inline prometheus/promhttp/http.go:404]:
```suggestion
	Enable bool
```

File: prometheus/promhttp/http.go
Comment (bwplotka) [inline prometheus/promhttp/http.go:395]:
```suggestion
	// Enable specifies if the experimental OpenMetrics encoding is added to the
```

File: prometheus/promhttp/http.go
Comment (bwplotka) [inline prometheus/promhttp/http.go:377]: Let's remove all the commentary above and leave only Deprecated note. Users can navigate to that field for the description.

File: prometheus/promhttp/http.go
Comment (bwplotka) [inline prometheus/promhttp/http.go:379]:
```suggestion
	// Deprecated: Use OpenMetricsOptions.Enable instead.
```

File: prometheus/promhttp/http.go
Comment (bwplotka) [inline prometheus/promhttp/http.go:403]:
```suggestion
	// possible options during content negotiation. 
	//
	// Note that Prometheus 2.5.0+ might negotiate OpenMetrics Text format 
	// as first priority unless user uses custom scrape protocol prioritization or
	// histograms feature is enabled (then Prometheus proto format is prioritized,
	// which client_golang supports).
	//
	// The main reason for enabling OpenMetrics Text was to enable exemplars,
	// but Prometheus proto has that as feature as well.
	// 
	// However, the move to OpenMetrics is not completely transparent. Most notably,
	// the values of "quantile" labels of Summaries and "le" labels of Histograms are
	// formatted with a trailing ".0" if they would otherwise look like integer numbers
	// (which changes the identity of the resulting series on the Prometheus
	// server).
	//
	// See other options in OpenMetricsOptions to learn how to enable some special
	// features e.g. potentially dangerous created timestamp series.
```

File: prometheus/promhttp/http.go
Comment (bwplotka) [inline prometheus/promhttp/http.go:419]:
```suggestion
	EnableCreatedSeries bool
```

File: prometheus/promhttp/http.go
Comment (bwplotka) [inline prometheus/promhttp/http.go:418]:
```suggestion
	// EnableCreatedSeries specifies if this handler should add, extra, synthetic
	// '_created' series for counters, histograms and summaries, as defined in
	// the OpenMetrics specification (see
	// https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#counter-1).
	// 
	// Created timestamps are used to improve the accuracy of reset detection,
	// but the way it's designed in OpenMetrics 1.0 it also dramatically increases cardinality
	// if the scraper does not handle those metrics correctly (converting to created timestamp
	// instead of leaving those series as-is). New OpenMetrics versions might improve
	// this situation.
	//
	// Prometheus introduced the feature flag 'created-timestamp-zero-ingestion'
	// in version 2.50.0, but only for the Prometheus protobuf format. Starting in 
	// future Prometheus version, the feature flag will be extended to the OpenMetrics
	// text format, thus safe to be enabled to improve accuracy of counters in Prometheus.
```

**Review by bwplotka — APPROVED** [review]
Epic, proposed some improvements, thanks!

File: prometheus/promhttp/http.go
Comment (ArthurSens) [reply prometheus/promhttp/http.go:403]:
```
	// The main reason for enabling OpenMetrics Text was to enable exemplars,
	// but Prometheus proto has that as feature as well.
```

this part seems a bit out of place, not adding much to the comment 🤔, is it ok to not add this part?

File: prometheus/promhttp/http.go
Comment (ArthurSens) [reply prometheus/promhttp/http.go:419]: Hmmmmm, future versions of OM might not expose them as a series. What do you think about `EnableCreatedTimestamps`?

**Comment by ArthurSens** [comment]
We need prometheus/common v0.49.0 for this feature, but this version of prometheus/common requires go 1.21.

Therefore, the PR is blocked until we drop support for go 1.20

**Review by kakkoyun — APPROVED** [review]
LGTM

**Comment by ArthurSens** [comment]
After quite some work to unblock this, this is finally ready :)

Do you want to review again or should I just merge?

**Comment by ArthurSens** [comment]
Now that https://github.com/prometheus/prometheus/pull/14356 was merged, should we revive this? :)

**Comment by ArthurSens** [comment]
PR rebased on main. Ready for another review just in case @kakkoyun @bwplotka @vesari

**Comment by ArthurSens** [comment]
Since we have approvals from the past, I'll take the liberty to merge this PR next week :)

File: prometheus/promhttp/http.go
Comment (bwplotka) [inline prometheus/promhttp/http.go:422]: Hm.. want to play with different ideas for this options first 🤔 Even keeping `EnableOpenMetrics` and then `OpenMetricsOptions.AddV1CreatedSeries` might a better choice? 🤔

**Review by bwplotka — CHANGES_REQUESTED** [review]
Let's timebox this, but wanted to try last more time to consider an improvement to the config structure, otherwise LGTM!

File: prometheus/promhttp/http.go
Comment (ArthurSens) [reply prometheus/promhttp/http.go:422]: > Hm.. want to play with different ideas for this options first 🤔 Even keeping `EnableOpenMetrics` and then `OpenMetricsOptions.AddV1CreatedSeries` might a better choice? 🤔

Interesting! I assume that this suggestion comes from not liking the idea of mentioning that "Created" is a "Series", this is something we'd like to change in the next version to something closer to metadata.

What do you think of `EnableCreatedTimestamps` instead? 

I think mentioning versions on fields is not the best, in the future I can see a `version` field that controls how each feature is exposed instead of `AddV1Created..`, `AddV2Created...` and then handling mutually exclusiveness

**Review by Maniktherana — COMMENTED** [review]
should we add summaries and counters in this example too?

File: prometheus/promhttp/http.go
Comment (ArthurSens) [reply prometheus/promhttp/http.go:422]: Oh lol, I just realized I suggested what's already in the code 🤦

File: prometheus/promhttp/http.go
Comment (bwplotka) [reply prometheus/promhttp/http.go]: If they are not expanded as series they are not a problem and will be fine by default (sorry for missing this commment)

File: prometheus/promhttp/http.go
Comment (bwplotka) [reply prometheus/promhttp/http.go:422]: I understand the confusion, here is why: https://github.com/prometheus/client_golang/pull/1408#discussion_r1774821420

We will likely not have "enable X Y Z" unless protocol will make a challenging decision like with CT being a series. So ideally we only have this "unsafe 1.0 Ct as series" flag for now, but do CT by default if they are safe to use (they don't increase 50-30% of Prometheus memory for no reason). (:

Comment (bwplotka) [reply prometheus/promhttp/http.go:422]: We might have some more flags in future for units too though, so some plan for those would be nice. For CT as series that an OpenMetrics option, but for units (e.g. AutogenerateUnits)... it's kind of general option? maybe not even in promhttp but in registry or something?

Since we now little and we already make it complex I would vote for keeping `EnableOpenMetrics` and adding `EmitOpenMetricsTextCreatedSamples bool` would be perfect for now. Samples not series as this is how it's referenced in OpenMetrics  https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#counter-1

Simple, and effective, then we probably should fail or ignore this if EnableOpenMetrics is not true.

WDYT

Comment (vesari) [reply prometheus/promhttp/http.go:422]: On a loosely relate note, I would agree on a general option for units. Happy to resume the work on that one whenever (if ever) needed.

**Comment by ArthurSens** [comment]
> should we add summaries and counters in the example too?

So far all our examples just add one or two metrics, not sure if that's needed 🤷

File: prometheus/promhttp/http.go
Comment (ArthurSens) [reply prometheus/promhttp/http.go:422]: I'm still confused here haha

`EmitOpenMetricsTextCreatedSamples bool`, this means there's no distinction between OM v1 and v2, do I understand that correctly? If I'm following you correctly, if we introduce this flag now with `false` as the default value, we won't be able to change it once OM v2 is out, right?

Comment (bwplotka) [reply prometheus/promhttp/http.go:422]: That's exactly why I call this option "CreatedSamples", so the not ideal _created lines. Once in V2 we have that in metadata or as annotation, even with this option false we will emit CT semantics by default IMO. 

This the same as with exemplars - enable open metrics, enables exemplars, no questions asked, because they were designed to NOT spam TSDB with odd series (:

Perhaps let's chat offline if this is still not clear/I am missing something? (:

**Comment by ArthurSens** [comment]
@bwplotka should be ready for review again!

File: prometheus/promhttp/http.go
Comment (krajorama) [inline prometheus/promhttp/http.go:431]: I feel that including `Emit` is a bit strong word because switching to OM v2 if  OM v2 has the metadata support for created timestamp AND someone set `EmitOpenMetricsTextCreatedSamples`  to true, they will no longer see the `_created` samples with 2.0 , only 1.0 - because we want to move CT into metadata. Which would mean we contradict the call to action `Emit!`.  Maybe calling it `EnableOpenMetricsTextCreatedSamples` would be slightly better to not contradict ourselves in 2.0.

Comment (bwplotka) [reply prometheus/promhttp/http.go:431]: +1 cc @ArthurSens WDYT?

Comment (ArthurSens) [reply prometheus/promhttp/http.go]: Yep, sounds better!

**Review by krajorama — APPROVED** [review]
SGTM

**Review by bwplotka — COMMENTED** [review]
LGTM, thanks!

**Review by bwplotka — APPROVED** [review]

**Pull request merged**
